external/esp_idf_port/esp32/esp32_queue_api.c

```diff
@@ -0,0 +1,434 @@
+ * Return:
+ *	Returns pdPASS if success, pdFAIL if failure.
+ *
+ ****************************************************************************/
+int32_t IRAM_ATTR queue_send_from_isr_wrapper(void *queue, void *item, void *hptw)
+{
+	int32_t ret;
+	queue_info_t *queue_info = NULL;
+
+	if (!queue || !item) {
+		return pdFAIL;
+	}
+	queue_info = (queue_info_t *) queue;
+	if (queue_info->mqd_fd_send != (mqd_t) ERROR) {
+		ret = mq_send(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, NORMAL);
+		if (ret == ERROR) {
+			ets_printf("queue_send_from_isr_wrapper error\n");
+			return pdFAIL;
+		}
+		return pdPASS;
+	}
+	return pdFAIL;
+}
+
+/****************************************************************************
+ * Name: queue_send_to_back_wrapper
+ *
+ * Description:
+ *	This function is the wrapper of queue_send_wrapper to adds the specificied message (msg) to
+ * the message queue (mqdes) end.
+ *
+ * Inputs:
+ *	queue - The queue to send messages.
+ *	item - The message to be send.
+ *	block_time_tick - the time (in Ticks) to wait until a timeout is decleared
+ * Return:
+ *	Returns pdPASS if success, pdFAIL if failure.
+ *
+ ****************************************************************************/
+int32_t IRAM_ATTR queue_send_to_back_wrapper(void *queue, void *item, uint32_t block_time_tick)
+{
+	if (!queue || !item) {
+		return pdFAIL;
+	}
+	return queue_send_wrapper(queue, item, block_time_tick);
+}
+
+/****************************************************************************
+ * Name: queue_send_to_front_wrapper
+ *
+ * Description:
+ *	This function is the wrapper of mq_timedsend to adds the specificied message (msg) to
+ * the message queue (mqdes) head.
+ *
+ * Inputs:
+ *	queue - The queue to send messages.
+ *	item - The message to be send.
+ *	block_time_tick - the time (in Ticks) to wait until a timeout is decleared
+ * Return:
+ *	Returns pdPASS if success, pdFAIL if failure.
+ *
+ ****************************************************************************/
+int32_t IRAM_ATTR queue_send_to_front_wrapper(void *queue, void *item, uint32_t block_time_tick)
+{
+	int32_t ret;
+	queue_info_t *queue_info = NULL;
+	struct timespec abstime;
+	uint32_t msecs;
+	uint32_t secs;
+	uint32_t nsecs;
+
+	if (!queue || !item) {
+		return pdFAIL;
+	}
+	queue_info = (queue_info_t *) queue;
+	if (queue_info->mqd_fd_send != (mqd_t) ERROR) {
+		if (block_time_tick == 0xFFFFFFFF) {
+			ret = mq_send(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, NORMAL);
+			if (ret == ERROR) {
+				return pdFAIL;
+			}
+		} else {
+			(void)clock_gettime(CLOCK_REALTIME, &abstime);
+			msecs = TICK2MSEC(block_time_tick);
+			secs = msecs / MSEC_PER_SEC;
+			nsecs = (msecs - (secs * MSEC_PER_SEC)) * NSEC_PER_MSEC;
+			abstime.tv_sec += secs;
+			abstime.tv_nsec += nsecs;
+			ret = mq_timedsend(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, HIGH, &abstime);
```

Comment:
by checking mq_send internal logic,  it can not gurantee 100% send_to_front.
it may have little influence,  and not need to fix it by now, but let's note a TODO here

Reply:
mark it.